# ============================================================
# Models comparsion
# ============================================================

In [4]:
import json
import pandas as pd

with open("logreg_results.json", "r") as f:
    lr_results = json.load(f)

with open("softmax_results.json", "r") as f:
    softmax_results = json.load(f)

with open("final_results_nn.json", "r") as f:
    nn_best_results = json.load(f)

with open("best_cnn_results.json", "r") as f:
    cnn_best_results = json.load(f)

df_results = pd.DataFrame([
    {'Model': 'Logistic Regression (Manual)', 'Test Accuracy (%)': lr_results["test_accuracy"] * 100, 'Test Loss': lr_results["test_loss"], 'Training Time (s)': lr_results["training_time_lr"]},
    {'Model': 'Softmax Regression (Manual)', 'Test Accuracy (%)': softmax_results["test_accuracy_manual"] * 100, 'Test Loss': softmax_results["test_loss_manual"], 'Training Time (s)': softmax_results["train_time_manual"]},
    {'Model': 'Softmax Regression (Built-in)', 'Test Accuracy (%)': softmax_results["test_accuracy_builtin"] * 100, 'Test Loss': softmax_results["test_loss_builtin"], 'Training Time (s)': softmax_results["training_time_builtin"]},
    {'Model': 'Neural Network (Best Params)', 'Test Accuracy (%)': nn_best_results["test_accuracy_percent"] , 'Test Loss': nn_best_results["test_loss"], 'Training Time (s)': nn_best_results["training_time_sec"]},
    {'Model': 'CNN (Best Model)', 'Test Accuracy (%)': cnn_best_results["test_accuracy_percent"] , 'Test Loss': cnn_best_results["test_loss"], 'Training Time (s)': cnn_best_results["training_time_sec"]}
])

display(df_results)

Model  Test Accuracy (%)  Test Loss  \
0   Logistic Regression (Manual)          99.763191   0.012224   
1    Softmax Regression (Manual)          90.764284   0.334960   
2  Softmax Regression (Built-in)          90.764284   0.334960   
3   Neural Network (Best Params)          97.800000   0.075800   
4               CNN (Best Model)          98.400000   0.049400   

   Training Time (s)  
0           0.922057  
1           8.376068  
2           0.000791  
3         198.970000  
4         465.710000

## Model Comparison: Computational Complexity & Training Time

| # | Model | Test Accuracy (%) | Test Loss | Training Time (s) |
|:-:|:-----------------------------|:----------------:|:------------:|:----------------:|
| 0 | Logistic Regression (Manual) | **99.76** | 0.012 | 0.92 |
| 1 | Softmax Regression (Manual)  | 90.76 | 0.335 | 8.38 |
| 2 | Softmax Regression (Built-in) | 90.76 | 0.335 | **0.0008** |
| 3 | Neural Network (Best Params) | 97.80 | 0.076 | 198.97 |
| 4 | CNN (Best Model) | 98.40 | **0.049** | **465.71** |

---

###  Computational Complexity Analysis

- **Logistic Regression (Manual):**
  - Complexity ≈ *O(n × d)* per iteration (where *n* = samples, *d* = features).
  - Extremely lightweight; no hidden layers.
  - Minimal training time and high accuracy because MNIST is linearly separable for most digits.

- **Softmax Regression:**
  - Adds multi-class capability (10 outputs).
  - Similar complexity to logistic regression but involves more matrix operations.
  - The **manual** version is slower due to Python loops and non-optimized operations.
  - The **built-in** version (e.g., using TensorFlow/PyTorch) is heavily optimized with vectorization and GPU acceleration — hence **massive speedup**.

- **Neural Network (Best Params):**
  - Complexity ≈ *O(L × n × d)* where *L* = layers.
  - Nonlinear activations and backpropagation increase cost.
  - More expressive model → better generalization but **higher training time**.

- **CNN (Best Model):**
  - Convolutional layers dominate: *O(n × k² × c_in × c_out × h × w)*.
  - Much more computationally heavy due to convolutions and pooling.
  - However, it captures spatial features, leading to **best loss** and **strong generalization** on image data.
  - Slowest to train but most powerful.

---

###  When to Use Each Model

| Model | When to Use |
|:------|:-------------|
| **Logistic Regression** | When data is simple, linearly separable, and you need ultra-fast inference or baselines. |
| **Softmax Regression** | For multi-class classification with moderate complexity and small datasets. |
| **Neural Network** | When relationships are nonlinear and you need more flexibility without going deep into CNNs. |
| **CNN** | When dealing with **image data** or any data with spatial/temporal structure — the best accuracy–complexity trade-off if compute allows. |

---

###  Summary
- **Fastest model:** Built-in Softmax Regression.  
- **Best accuracy:** Logistic Regression surprisingly tops due to MNIST’s simple structure.  
- **Best model for real-world image tasks:** CNN.  
- **Best balance for scalability:** Neural Network (Best Params).


